### DISCUSSION TITLE
Why isn't `ColumnInfo.columnType` public?

### DISCUSSION THREAD

**Discussion by mallman** [groue/GRDB.swift#1560] [Q&A] [answered]
Why isn't `ColumnInfo.columnType` public? Its type (`Database.ColumnType`) is public. It provides a safer alternative to `ColumnInfo.type`, which is public.

I want to convert the value of a `DatabasePreUpdateEvent` column to a Swift `Bool` if the table column is declared as a boolean. I can check if `columnInfo.type.lowercased() == "boolean"`, but checking `columnInfo.column == .boolean` would be much cleaner.

Also, as my ultimate goal is to get the value of a column as a `Bool` in introspection if the table column type is `BOOLEAN`, is there a GRDB API for that already? Right now I'm using `DatabaseValue.Storage.value` to get the column value, but that is based on the storage type, not the declared type (like boolean).

Thank you!

**Comment by groue** [comment, marked as answer]
Hello @mallman,

> Why isn't `ColumnInfo.columnType` public?

This is because `ColumnType` is a type dedicated to schema definition, not to schema introspection.

`ColumnType` provides ready-made constants for frequently used column types: `.integer`, `.text`, `.boolean`, etc. But it is not an enum. It is a struct. SQLite accepts about everything as a column type. For example, consider those three columns:

```swift
try db.create(table: "myTable") { t in
  t.column("a", .boolean)                        // 1
  t.column("b", ColumnType(rawValue: "BOOL"))    // 2
  t.column("c", ColumnType(rawValue: "PIKACHU")) // 3
}
```

All of them are valid ways to declare columns that store booleans. 2 is original but quite OK; 3 is confusing but it won't create any runtime problem at all. That's how [Datatypes In SQLite](https://www.sqlite.org/datatype3.html) work.

The library does not provide any interpretation of raw column types returned by `ColumnInfo.type` (i.e. strings like "BOOLEAN", "BOOL", "PIKACHU").

If an _application_ wants to perform a specific handling of columns with type "BOOLEAN", that's ok: `columnInfo.type.lowercased() == "boolean"`. It can also perform a specific handling of column types "BOOL" and "PIKACHU" if it wants. That's an application concern.

To my eyes, the `columnInfo.columnType == .boolean` expression looks extremely misleading, because it looks like it correctly detects all columns that store bools. But no, it only detects columns declared with the "BOOLEAN" type.

**Reply by groue** [reply]
> Also, as my ultimate goal is to get the value of a column as a `Bool` in introspection if the table column type is `BOOLEAN`, is there a GRDB API for that already?

There exists [`sqlite3_column_decltype`](https://www.sqlite.org/c3ref/column_decltype.html):

```swift
try db.create(table: "myTable") { t in
    t.column("flag", .boolean)
}

let sql = "SELECT flag from myTable"
let statement = try db.makeStatement(sql: sql)
if let type = sqlite3_column_decltype(statement.sqliteStatement, 0) {
    print(String(cString: type))
} else {
    print("unknown")
}
```

The above snippet prints "BOOLEAN".

Add the NOT operator: "SELECT NOT flag from myTable". Now it prints "unknown".

So it won't bring us very far.

> Right now I'm using `DatabaseValue.Storage.value` to get the column value, but that is based on the storage type, not the declared type (like boolean).

Yes. We can't do much more with the information provided by SQLite.

Instead of switching on the storage of a `DatabaseValue`, you can use `fromDatabaseValue`:

```swift
if dbValue.isNull {
  // Handle NULL
} else if let bool = Bool.fromDatabaseValue(dbValue) {
  // We were able to decode a bool, but that's all that we know.
} else {
  // Definitely not a bool
}
```

**Comment by mallman** [comment]
Hi @groue. I believe I understand your reply, and it makes sense. `ColumnInfo.columnType` provides standardized values for schema definition with GRDB, but it does not represent any kind of standard in SQLite.

Rather than introspecting on the SQLite column info type, how about introspecting on the GRDB migration schema? As a matter of API the user designs and defines the migrations. They provide a column type annotation of their choosing. It can be one of the convenient pre-defined values, or it can be "PIKACHU", et. al. The point is, they control what they put in there. They should be able to identify it when introspecting a table migration.

Is it even possible to introspect on the schema definition? I don't see it stored in the table, and defining a stable serialized form for it may be a major pain in the ass.

Another possibility is I can probably introspect on the Swift type of the `TableRecord` property.

At any rate, for my purposes, assuming the value of `ColumnInfo.type` is "BOOLEAN" will work.

**Reply by groue** [reply]
> Hi @groue. I believe I understand your reply, and it makes sense. `ColumnInfo.columnType` provides standardized values for schema definition with GRDB, but it does not represent any kind of standard in SQLite.

Basically yes. On top of that, add the first sentence of the GRBD documentation for the [database schema](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databaseschema), written in bold:

> **GRDB supports all database schemas, and has no requirement.**

[🎵 _Come as you are_ 🙂](https://www.youtube.com/watch?v=W2QeQ9ZufAk). I care a lot about this. Some people ship ready-made databases as resources in their app, and god knows where the db comes from 😅.

The same document goes on with some recommendations, though, as few as it was possible. Each one of them has a rationale, and a way to deal with deviations (so that _really_ all schemas are supported).

The SQL type of boolean columns is not mentioned because there is no GRDB feature that depends on it.

> Rather than introspecting on the SQLite column info type, how about introspecting on the GRDB migration schema? As a matter of API the user designs and defines the migrations. They provide a column type annotation of their choosing. It can be one of the convenient pre-defined values, or it can be "PIKACHU", et. al. The point is, they control what they put in there. They should be able to identify it when introspecting a table migration.
> 
> Is it even possible to introspect on the schema definition? I don't see it stored in the table, and defining a stable serialized form for it may be a major pain in the ass.

The only serialization of migrations I know is the Swift files that define them 😉 Maybe SPM plugins and Swift syntax can do miracles. For an example of a Swift SQLite package that uses plugins for introspection, see [Lighter](https://github.com/Lighter-swift/Lighter).

At runtime, an app can use [the schema introspection methods](https://swiftpackageindex.com/groue/grdb.swift/documentation/grdb/databaseschema#querying-the-database-schema) which provide as much information as SQLite provides (i.e. strings).

An app can also parse the results of `SELECT * FROM sqlite_master`. GRDB does not ship with any SQL parser, and I don't want that it ships one, because it will create issues as soon as SQLite ships a new version and new SQL features.

> Another possibility is I can probably introspect on the Swift type of the `TableRecord` property.
> 
> At any rate, for my purposes, assuming the value of `ColumnInfo.type` is "BOOLEAN" will work.

Yes. I think that you can **decide** that your app only works with `BOOLEAN` columns, i.e. it will fail to deal with booleans when the column type is `BOOL` or `PIKACHU`. If you control the database, it is easy. If you have to deal with wild databases, then you can document that boolean columns must be declared as `BOOLEAN` 😉
